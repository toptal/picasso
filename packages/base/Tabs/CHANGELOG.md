# @toptal/picasso-tabs

## 2.0.3

### Patch Changes

- Updated dependencies [[`978ae36`](https://github.com/toptal/picasso/commit/978ae36fa8e0306ecbffdddf1a725dc0997c3d6e)]:
  - @toptal/picasso-icons@1.2.0
  - @toptal/picasso-user-badge@2.0.3

## 2.0.2

### Patch Changes

- Updated dependencies [[`d599529`](https://github.com/toptal/picasso/commit/d599529bcb283c367b63c612fee81394e66c9740)]:
  - @toptal/picasso-shared@15.0.0
  - @toptal/picasso-utils@1.0.3
  - @toptal/picasso-user-badge@2.0.2
  - @toptal/picasso-container@1.0.3
  - @toptal/picasso-typography@2.0.1
  - @toptal/picasso-typography-overflow@2.0.1
  - @toptal/picasso-icons@1.1.1

## 2.0.1

### Patch Changes

- Updated dependencies [[`c7560ae`](https://github.com/toptal/picasso/commit/c7560aed9dd41bb458c5532608ddd542890523e5)]:
  - @toptal/picasso-icons@1.1.0
  - @toptal/picasso-user-badge@2.0.1

## 2.0.0

### Major Changes

- [#4200](https://github.com/toptal/picasso/pull/4200) [`4ee1ebdafd9e5830d5ec6007620186d5a61befee`](https://github.com/toptal/picasso/commit/4ee1ebdafd9e5830d5ec6007620186d5a61befee) Thanks [@mkrl](https://github.com/mkrl)!
- added a `tailwindcss` dependency (tailwindcss is now required to make this package work)

### Patch Changes

- Updated dependencies [[`4ee1ebd`](https://github.com/toptal/picasso/commit/4ee1ebdafd9e5830d5ec6007620186d5a61befee), [`4ee1ebd`](https://github.com/toptal/picasso/commit/4ee1ebdafd9e5830d5ec6007620186d5a61befee)]:
  - @toptal/picasso-typography@2.0.0
  - @toptal/picasso-typography-overflow@2.0.0
  - @toptal/picasso-user-badge@2.0.0

## 1.0.5

### Patch Changes

- [#4224](https://github.com/toptal/picasso/pull/4224) [`c05387de5636ed094365d1eff67b955d84b81c61`](https://github.com/toptal/picasso/commit/c05387de5636ed094365d1eff67b955d84b81c61) Thanks [@TomasSlama](https://github.com/TomasSlama)!
- revert types for compound components

- Updated dependencies []:
  - @toptal/picasso-user-badge@1.0.4

## 1.0.4

### Patch Changes

- [#4216](https://github.com/toptal/picasso/pull/4216) [`3512588b06c3660471a68500275321c640278cf0`](https://github.com/toptal/picasso/commit/3512588b06c3660471a68500275321c640278cf0) Thanks [@TomasSlama](https://github.com/TomasSlama)!
- update types for the compound component

- Updated dependencies []:
  - @toptal/picasso-user-badge@1.0.3

## 1.0.3

### Patch Changes

- Updated dependencies []:
  - @toptal/picasso-typography-overflow@1.0.3

## 1.0.2

### Patch Changes

- Updated dependencies [[`a1d523092cc4de6cb376156435b99b1e483f39b9`](https://github.com/toptal/picasso/commit/a1d523092cc4de6cb376156435b99b1e483f39b9)]:
  - @toptal/picasso-shared@14.0.1
  - @toptal/picasso-utils@1.0.2
  - @toptal/picasso-user-badge@1.0.2
  - @toptal/picasso-container@1.0.2
  - @toptal/picasso-typography@1.0.2
  - @toptal/picasso-typography-overflow@1.0.2
  - @toptal/picasso-icons@1.0.2

## 1.0.1

### Patch Changes

- [#4164](https://github.com/toptal/picasso/pull/4164) [`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30) Thanks [@mkrl](https://github.com/mkrl)!
- migrate to Picasso by Parts
  - picasso is now distributed as a set of independent packages with the main package `@toptal/picasso` now being a collection of re-exported packages
- Updated dependencies [[`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30)]:
  - @toptal/picasso-shared@14.0.0
  - @toptal/picasso-typography-overflow@1.0.1
  - @toptal/picasso-typography@1.0.1
  - @toptal/picasso-container@1.0.1
  - @toptal/picasso-user-badge@1.0.1
  - @toptal/picasso-icons@1.0.1
  - @toptal/picasso-utils@1.0.1
