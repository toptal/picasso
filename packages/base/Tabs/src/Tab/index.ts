import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Tab'

export { default as Tab } from './Tab'
export type TabProps = OmitInternalProps<Props>
