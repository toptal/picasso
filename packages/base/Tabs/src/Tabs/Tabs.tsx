import type { ReactNode } from 'react'
import React, { forwardRef } from 'react'
import type { Theme } from '@material-ui/core/styles'
import { makeStyles } from '@material-ui/core/styles'
import type { TabsProps } from '@material-ui/core'
import { Tabs as MUITabs } from '@material-ui/core'
import type { ButtonOrAnchorProps, BaseProps } from '@toptal/picasso-shared'

import TabScrollButton from '../TabScrollButton'
import styles from './styles'
import useTabAction from './use-tab-action'

export interface Props
  extends BaseProps,
    Omit<ButtonOrAnchorProps, 'onChange'> {
  /** Tabs content containing Tab components */
  children: ReactNode

  /** Callback fired when the value changes. */
  onChange?: (event: React.ChangeEvent<{}>, value: TabsProps['value']) => void

  /** The value of the currently selected Tab. If you don't want any selected Tab, you can set this property to false. */
  value: TabsProps['value']

  /** The tabs orientation (layout flow direction). */
  orientation?: 'horizontal' | 'vertical'

  /** Determines additional display behavior of the tabs */
  variant?: Extract<TabsProps['variant'], 'scrollable' | 'fullWidth'>
}

const useStyles = makeStyles<Theme>(styles, {
  name: 'Tabs',
})

export const TabsOrientationContext = React.createContext<
  'horizontal' | 'vertical'
>('horizontal')

// eslint-disable-next-line react/display-name
export const Tabs = forwardRef<HTMLButtonElement, Props>(function Tabs(
  props,
  ref
) {
  const {
    children,
    orientation,
    onChange,
    value,
    variant = 'scrollable',
    ...rest
  } = props
  const classes = useStyles(props)
  const action = useTabAction()

  return (
    <TabsOrientationContext.Provider value={orientation!}>
      <MUITabs
        {...rest}
        classes={{ root: classes[orientation!] }}
        ref={ref}
        onChange={onChange}
        value={value}
        action={action}
        scrollButtons='auto'
        ScrollButtonComponent={TabScrollButton}
        orientation={orientation}
        variant={variant}
      >
        {children}
      </MUITabs>
    </TabsOrientationContext.Provider>
  )
})

Tabs.displayName = 'Tabs'
Tabs.defaultProps = {
  orientation: 'horizontal',
}

export default Tabs
