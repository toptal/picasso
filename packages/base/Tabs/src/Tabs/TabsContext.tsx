import React from 'react'

export interface TabsContextValue<T> {
  value: T
  onChange: (event: React.ChangeEvent<{}>, value: T) => void
  orientation: 'horizontal' | 'vertical'
  variant: 'scrollable' | 'fullWidth'
  direction?: 'ltr' | 'rtl'
}

export const TabsContext = React.createContext<TabsContextValue<any>>({
  value: undefined,
  onChange: () => {},
  orientation: 'horizontal',
  variant: 'scrollable',
  direction: 'ltr',
})

export const useTabsContext = () => {
  const context = React.useContext(TabsContext)

  if (!context) {
    throw new Error('useTabsContext must be used within a TabsProvider')
  }

  return context
}
