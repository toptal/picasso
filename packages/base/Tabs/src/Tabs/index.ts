import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props, TabsValueType } from './Tabs'

export { default as Tabs } from './Tabs'
export type TabsProps = OmitInternalProps<Props<TabsValueType>>
export type { TabsValueType } from './Tabs'
