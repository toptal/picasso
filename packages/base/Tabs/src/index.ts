export * from './Tabs'
export * from './Tab'
export * from './TabDescription'
export * from './TabsCompound'
export * from './TabLabel'
