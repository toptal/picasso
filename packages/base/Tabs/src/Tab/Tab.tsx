import type { ReactNode, HTMLAttributes, ReactElement } from 'react'
import React, { forwardRef, useContext } from 'react'
import type { TabProps } from '@mui/base/Tab'
import { Tab as MUITab } from '@mui/base/Tab'
import type { BaseProps, TextLabelProps } from '@toptal/picasso-shared'
import { useTitleCase } from '@toptal/picasso-shared'
import { UserBadge } from '@toptal/picasso-user-badge'
import { twJoin, twMerge } from '@toptal/picasso-tailwind-merge'

import { TabsContext } from '../Tabs/Tabs'
import { TabLabel } from '../TabLabel'
import { TabDescription } from '../TabDescription'

export interface Props
  extends BaseProps,
    TextLabelProps,
    Omit<HTMLAttributes<HTMLButtonElement>, 'onChange'> {
  /**
   * If true, the tab will be disabled
   * @default false
   */
  disabled?: boolean

  /** You can provide your own value. Otherwise, we fallback to the child position index */
  value?: TabProps['value']

  /** The label element */
  label?: ReactNode

  /** The Icon element */
  icon?: ReactElement

  /** Image URL */
  avatar?: string | null

  /** Description */
  description?: string

  // Properties below are managed by Tabs component

  selected?: boolean
  onChange?: TabProps['onChange']
  onClick?: TabProps['onClick']
}

const getOpacityClass = (
  selected: boolean,
  disabled: boolean,
  orientation: 'horizontal' | 'vertical'
) => {
  if (disabled) {
    return 'opacity-50'
  }

  if (selected || orientation === 'vertical') {
    return 'opacity-100 '
  }

  return 'opacity-70'
}

const wrapperClassesByOrientation = {
  horizontal: 'inline-flex items-center flex-col justify-center',
  vertical: 'block',
}

const rootClassesByOrientation = (selected: boolean) => ({
  horizontal: [
    'm-0 [&:not(:last-child)]:mr-8 pt-[0.5625rem] pb-[0.4375rem] px-0',
    'text-center bg-transparent transition-shadow z-10 rounded-none',
    selected ? 'text-black' : 'text-inheritColor',
    selected && 'shadow-blue-500 shadow-[inset_0_-2px_0]',
  ],
  vertical: [
    ' first:mt-4 last:mb-4 my-1 mx-0 py-2 px-4',
    'text-left rounded-l-sm rounded-r-none transition-all',
    'w-full overflow-hidden',
    selected && 'shadow-1',
    selected && [
      'before:absolute',
      'before:content-[""]',
      'before:bottom-0',
      'before:left-0',
      'before:top-0',
      'before:w-[3px]',
      'before:bg-blue-500',
    ],
    selected
      ? 'bg-gray-50 text-black'
      : 'bg-gray-100 hover:bg-gray-200 text-graphite-700 hover:text-black',
  ],
})

const classesByVariant = {
  scrollable: 'shrink-0 max-w-[264px]',
  fullWidth: 'shrink flex-grow basis-0',
}

export const Tab = forwardRef<HTMLButtonElement, Props>(function Tab(
  props,
  ref
) {
  const {
    disabled,
    value,
    label,
    icon,
    onChange,
    onClick,
    titleCase: propsTitleCase,
    description,
    avatar,
    className,
    ...rest
  } = props
  const titleCase = useTitleCase(propsTitleCase)
  const { orientation, variant } = useContext(TabsContext)
  const isHorizontal = orientation === 'horizontal'

  const renderLabel = getLabelComponent({
    avatar,
    description,
    disabled,
    label,
    orientation,
    titleCase,
  })

  return (
    <MUITab
      className=''
      {...rest}
      ref={ref}
      tabIndex={0}
      disabled={disabled}
      value={value}
      onChange={onChange}
      onClick={onClick}
      slotProps={{
        root: ownerState => {
          return {
            className: twMerge(
              getOpacityClass(
                ownerState.selected,
                ownerState.disabled,
                orientation
              ),
              rootClassesByOrientation(ownerState.selected)[orientation],
              classesByVariant[variant],
              ownerState.disabled
                ? 'cursor-default text-gray-500'
                : 'cursor-pointer',
              ownerState.disabled && 'pointer-events-none',
              icon && isHorizontal && 'min-h-0 pt-[0.5625rem] pr-6',
              'min-w-0 sm:min-w-[160px] md:min-w-[auto]',
              'border-0 cursor-pointer inline-flex outline-none',
              'items-center select-none align-middle appearance-none',
              'justify-center no-underline [-webkit-tap-highlight-color:transparent]',
              'normal-case whitespace-normal leading-4',
              'relative ',
              className
            ),
          }
        },
      }}
    >
      <span
        className={twJoin('w-full', wrapperClassesByOrientation[orientation])}
      >
        {renderLabel}
        {icon && <span className='absolute right-0 mb-0 h-4'>{icon}</span>}
      </span>
    </MUITab>
  )
})

Tab.defaultProps = {}

Tab.displayName = 'Tab'

type GetLabelComponentProps = {
  avatar?: string | null
  description?: string
  disabled?: boolean
  label?: React.ReactNode
  orientation: 'horizontal' | 'vertical'
  titleCase?: boolean
}
const getLabelComponent = ({
  avatar,
  description,
  disabled,
  label,
  orientation,
  titleCase,
}: GetLabelComponentProps): React.ReactNode => {
  if (!label) {
    return null
  }

  const isHorizontal = orientation === 'horizontal'
  const isCustomLabel = typeof label !== 'string'

  const Label = () => (
    <TabLabel titleCase={titleCase} label={label} orientation={orientation} />
  )

  if (isHorizontal || isCustomLabel) {
    return <Label />
  }

  if (typeof avatar === 'undefined') {
    return (
      <>
        <Label />
        {description && (
          <TabDescription disabled={disabled}>{description}</TabDescription>
        )}
      </>
    )
  }

  return (
    <UserBadge renderName={Label} name={label} avatar={avatar}>
      {description && (
        <TabDescription disabled={disabled}>{description}</TabDescription>
      )}
    </UserBadge>
  )
}

export default Tab
