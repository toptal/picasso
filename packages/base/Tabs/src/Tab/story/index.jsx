import { Tab } from '../Tab'
import PicassoBook from '~/.storybook/components/PicassoBook'

const componentDocs = PicassoBook.createComponentDocs(Tab, 'Tabs.Tab')

const chapter = PicassoBook.connectToPage(page => {
  page
    .createChapter('Tabs.Tab')
    .addExample(
      'Tab/story/Vertical.example.tsx',
      {
        title: 'Vertical tab',
        takeScreenshot: false,
      },
      'base/Tabs'
    )
    .addExample('Tab/story/Disabled.example.tsx', 'Disabled tab', 'base/Tabs')
    .addExample(
      'Tab/story/CustomValue.example.tsx',
      'Using custom value',
      'base/Tabs'
    )
    .addExample(
      'Tab/story/IconOrBadge.example.tsx',
      'With Icon or Badge',
      'base/Tabs'
    )
})

export default {
  chapter,
  componentDocs,
}
