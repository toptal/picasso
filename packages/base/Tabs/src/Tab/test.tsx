import React from 'react'
import { render } from '@toptal/picasso-test-utils'
import type { OmitInternalProps } from '@toptal/picasso-shared'
import * as titleCaseModule from 'ap-style-title-case'
import type { PicassoConfig } from '@toptal/picasso-test-utils'

import type { Props } from './Tab'
import { Tab } from './Tab'
import { Tabs } from '../Tabs'

jest.mock('ap-style-title-case')

const renderTab = (
  props: OmitInternalProps<Props>,
  picassoConfig?: PicassoConfig
) => {
  const { label, disabled, icon, titleCase } = props

  return render(
    <Tabs value={null}>
      <Tab
        label={label}
        disabled={disabled}
        icon={icon}
        titleCase={titleCase}
      />
    </Tabs>,
    undefined,
    picassoConfig
  )
}

let spiedOnTitleCase: jest.SpyInstance

describe('Tab', () => {
  beforeEach(() => {
    spiedOnTitleCase = jest.spyOn(titleCaseModule, 'default')
  })
  afterEach(() => {
    spiedOnTitleCase.mockReset()
  })

  describe('Tab', () => {
    it('renders', () => {
      const { container } = renderTab({
        label: 'Tab Label',
      })

      expect(container).toMatchSnapshot()
    })

    it('disabled tab', () => {
      const { container } = renderTab({
        label: 'Tab Label',
        disabled: true,
      })

      expect(container).toMatchSnapshot()
    })

    it('tab with icon', () => {
      const Icon = () => <div id='Icon' />
      const { container } = renderTab({
        label: 'Tab Label',
        icon: <Icon />,
      })

      expect(container).toMatchSnapshot()
    })

    it('should transform text to title case when Picasso titleCase property is true', () => {
      const LABEL_TEXT = 'Test vh2'

      renderTab({ label: LABEL_TEXT }, { titleCase: true })

      expect(spiedOnTitleCase).toHaveBeenCalledWith(LABEL_TEXT)
    })

    it('should not transform text to title case when Picasso titleCase property is true but the component property overrides it', () => {
      renderTab({ label: 'abc sp3', titleCase: false }, { titleCase: true })

      expect(spiedOnTitleCase).toHaveBeenCalledTimes(0)
    })
  })
})
