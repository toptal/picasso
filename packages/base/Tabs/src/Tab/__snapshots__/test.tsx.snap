// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab Tab disabled tab 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiTabs-root Tabs-horizontal"
    >
      <div
        class="MuiTabs-scrollable"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-scrollable"
        style="margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root PicassoTab-horizontal MuiTab-textColorInherit Mui-disabled Mui-disabled"
            disabled=""
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiTab-wrapper PicassoTab-wrapper"
            >
              <div
                class="m-0 text-sm text-inherit font-semibold"
              >
                Tab Label
              </div>
            </span>
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root PrivateTabIndicator-colorSecondary MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Tab Tab renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiTabs-root Tabs-horizontal"
    >
      <div
        class="MuiTabs-scrollable"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-scrollable"
        style="margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root PicassoTab-horizontal MuiTab-textColorInherit"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiTab-wrapper PicassoTab-wrapper"
            >
              <div
                class="m-0 text-sm text-inherit font-semibold"
              >
                Tab Label
              </div>
            </span>
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root PrivateTabIndicator-colorSecondary MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Tab Tab tab with icon 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiTabs-root Tabs-horizontal"
    >
      <div
        class="MuiTabs-scrollable"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-scrollable"
        style="margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root PicassoTab-horizontal MuiTab-textColorInherit MuiTab-labelIcon"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiTab-wrapper PicassoTab-wrapper"
            >
              <div
                id="Icon"
              />
              <div
                class="m-0 text-sm text-inherit font-semibold"
              >
                Tab Label
              </div>
            </span>
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root PrivateTabIndicator-colorSecondary MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;
